src/properties/property-effects.html

```diff
@@ -77,21 +82,59 @@
   // -- effects ----------------------------------------------
 
   /**
-   * Runs all effects for the given property on an instance.
+   * Runs all effects of a given type for the given set of property changes
+   * on an instance.
    *
    * @param {Object} inst The instance with effects to run
-   * @param {string} property Name of property
-   * @param {*} value Current value of property
-   * @param {*} old Previous value of property
-   * @param {Object<string,Array>} effects List of effects
+   * @param {string} type Type of effect to run
+   * @param {Object} props Bag of current property changes
+   * @param {Object} oldProps Bag of previous values for changed properties
    * @private
    */
-  function runEffects(inst, property, value, old, effects) {
-    for (let i=0, l=effects.length, fx; (i<l) && (fx=effects[i]); i++) {
-      if (Polymer.Path.matches(fx.path, property)) {
-        fx.fn(inst, property, inst.__data[property], old, fx.info);
+  function runEffects(inst, type, props, oldProps) {
+    let ran;
+    let effects = inst[type];
+    if (effects) {
+      let id = effectUid++;
+      for (let prop in props) {
+        if (runEffectsForProperty(inst, effects, id, prop, props[prop],
+          oldProps && oldProps[prop])) {
+          ran = true;
+        }
       }
     }
+    return ran;
+  }
+
+  /**
+   * Runs a list of effects for a given property.
+   *
+   * @param {Object} inst The instance with effects to run
+   * @param {Array} effects Array of effects
+   * @param {number} id Effect run id used for de-duping effects
+   * @param {string} prop Name of changed property
+   * @param {*} value Value of changed property
+   * @param {*} old Previous value of changed property
+   * @private
+   */
+  function runEffectsForProperty(inst, effects, id, prop, value, old) {
+    let ran;
+    let rootProperty = Polymer.Path.root(prop);
+    let fxs = effects[rootProperty];
+    if (fxs) {
+      let fromAbove = inst.__dataFromAbove;
+      for (let i=0, l=fxs.length, fx; (i<l) && (fx=fxs[i]); i++) {
+        if (Polymer.Path.matches(fx.path, prop) &&
+          (!fx.info || fx.info.lastRun !== id)) {
+          fx.fn(inst, prop, Polymer.Path.get(inst, prop) || value, old, fx.info, fromAbove);
```

Comment:
The `Polymer.Path.get` call here is a change from previous implementation which just did `inst.__data[prop]`, because paths are no longer persisted in `__data`; using `get` is more correct (better immunity against changes to path in earlier effect run this turn), but `inst.__data[prop] || value` passes all tests... revert?